tests/semantic_analyser.cpp

```diff
@@ -8,237 +11,150 @@
-}
-
-ast::ASTContext test(BPFtrace &bpftrace,
-                     const std::string &input,
-                     int expected_result,
-                     bool safe_mode = true)
-{
-  // This function will eventually be deprecated in favour of test_error()
-  assert(expected_result != 0 &&
-         "Use test(BPFtrace&, const std::string&) for expected successes");
-  return test(bpftrace, true, input, expected_result, {}, safe_mode, false);
+  return std::nullopt;
 }
 
-ast::ASTContext test(MockBPFfeature &feature, const std::string &input)
+template <typename T>
+std::optional<T> extract()
 {
-  auto bpftrace = get_mock_bpftrace();
-  bool mock_has_features = feature.has_features_;
-  return test(*bpftrace, mock_has_features, input, 0, {}, true, false);
-}
-
-ast::ASTContext test(MockBPFfeature &feature,
-                     const std::string &input,
-                     int expected_result,
-                     bool safe_mode = true)
-{
-  // This function will eventually be deprecated in favour of test_error()
-  assert(
-      expected_result != 0 &&
-      "Use test(MockBPFfeature&, const std::string&) for expected successes");
-  auto bpftrace = get_mock_bpftrace();
-  bool mock_has_features = feature.has_features_;
-  return test(*bpftrace,
-              mock_has_features,
-              input,
-              expected_result,
-              {},
-              safe_mode,
-              false);
+  return std::nullopt;
 }
 
-ast::ASTContext test(const std::string &input,
-                     int expected_result,
-                     bool safe_mode,
-                     bool has_child = false)
+std::string_view clean_prefix(std::string_view view)
 {
-  auto bpftrace = get_mock_bpftrace();
-  return test(
-      *bpftrace, true, input, expected_result, {}, safe_mode, has_child);
+  while (!view.empty() && view[0] == '\n')
+    view.remove_prefix(1); // Remove initial '\n'
+  return view;
 }
 
-ast::ASTContext test(const std::string &input, int expected_result)
-{
-  // This function will eventually be deprecated in favour of test_error()
-  assert(expected_result != 0 &&
-         "Use test(const std::string&) for expected successes");
-  auto bpftrace = get_mock_bpftrace();
-  return test(*bpftrace, true, input, expected_result, {}, true, false);
-}
+// This exists as a test fixture because the types may refer to `bpftrace`, so
+// this objects lifetime must exceed the tests lifetime. This is easier with a
+// fixture, and allows us to have a single harness.
+class SemanticAnalyserHarness {
+public:
+  template <Options options = Options{}, typename... Ts>
+    requires((std::is_same_v<std::decay_t<Ts>, Mock> ||
+              std::is_same_v<std::decay_t<Ts>, Warning> ||
+              std::is_same_v<std::decay_t<Ts>, NoWarning> ||
+              std::is_same_v<std::decay_t<Ts>, Error> ||
+              std::is_same_v<std::decay_t<Ts>, ExpectedAST>) &&
+             ...)
```

Comment:
This looks neat although I'm not much familiar with the syntax.

Do I understand correctly that this places restrictions on the types of values passed to `args` and if a different type is passed, the compiler will complain?

Also, what's the `&& ...` part for?

Reply:
Right, if you try to pass something else, then it will not compile. `std::decay_t` is just removing any qualifiers on the passed type (e.g. `const`, references, etc.).

The `&& ...` is a [fold expression](https://en.cppreference.com/w/cpp/language/fold.html) over the [parameter pack](https://en.cppreference.com/w/cpp/language/parameter_pack.html) Ts. Basically it is just requires that same expression `&&`ed for all types in `Ts`.